### PR TITLE
Alter menu items to use the new link parts

### PR DESCRIPTION
**Pull request by Mehdi-Miah**
- This change will utilise the new linkdestination and linkbehaviour parts

### CODE REVIEW

**Review by peterkeating — CHANGES_REQUESTED** [review]
This change doesn't currently cater for existing menu items on a site, which means if someone upgrades this module on an existing site then their menu will no longer work and they would have to manually update all the links.

**Comment by Mehdi-Miah** [comment]
@peterkeating hmm I didn't think of that, How do you think i should cater for both? 

possibly a conditional to check if it has the newly created parts or the older common menu item part.

**Comment by peterkeating** [comment]
> @peterkeating hmm I didn't think of that, How do you think i should cater for both?
> 
> possibly a conditional to check if it has the newly created parts or the older common menu item part.

I mentioned it when we briefed the changes. Because it's highly likely there aren't many menu items, I suggest we fetch the content items and update them to use the new parts.

**Comment by Mehdi-Miah** [comment]
would like to get your comments on this

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            var contentItems = await _session.Query<ContentItem>()
+                .With<ContentItemIndex>()
+                    .Where(x => x.Latest && x.ContentType == "Menu")
+                .ListAsync();
+
+            foreach (var contentItem in contentItems)
+            {
+                var menuItemsListPart = contentItem.As<MenuItemsListPart>();
+
+                foreach (var linkMenuItem in menuItemsListPart.MenuItems)
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:59]: I don't like how long this method is, let's break it up by moving the migration of menu items to it's own method that loops over the menu items and converts them.

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+                    try
+                    {
+                        if (linkMenuItem.As<CommonMenuItem>().ExternalUrl.Text != null)
+                        {
+                            linkDestinationPart = new LinkDestinationPart { DestinationUrl = linkMenuItem.As<CommonMenuItem>().ExternalUrl }; ;
+                        }
+                        else
+                        {
+                            var host = _httpContextAccessor.HttpContext.Request.Host.Value;
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:71]: What's going on here? We have had discussions offline about relative/absolute URLs and the fact it **must** be a relative URL otherwise moving content becomes a nightmare. Change this to be relative.

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                            var path = _httpContextAccessor.HttpContext.Request.Path.Value != null ? _httpContextAccessor.HttpContext.Request.Path.Value.Replace("admin","") : "/" ;
+
+                            linkDestinationPart = new LinkDestinationPart
+                            {
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:75]: Why's it checking whether the path has "admin" within it?

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                        _logger.LogError(ex, $"Failed to update menu item: {linkMenuItem.ContentType} ID : {linkMenuItem.ContentItemId}");
+                        continue;
+                    }
+                }
+
+                contentItem.Apply(nameof(MenuItemsListPart), menuItemsListPart);
+                await _contentManager.UpdateAsync(contentItem);
+                await _contentManager.PublishAsync(contentItem);
+            };
+
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:101]: Remove empty line.

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        }
+
+        public async Task<string> ContentItemIdToUrl(ContentItem linkMenuItem)
+        {
+            try
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:106]: Remove try/catch, it'll be handled within the main method on line 90.

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                }
+
+                contentItem.Apply(nameof(MenuItemsListPart), menuItemsListPart);
+                await _contentManager.UpdateAsync(contentItem);
+                await _contentManager.PublishAsync(contentItem);
+            };
+
+        }
+
+        public async Task<string> ContentItemIdToUrl(ContentItem linkMenuItem)
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:104]: Change method argument to a string that represents the `contentItemId`. There is no point passing in the whole content item because you're only using the content item ID. Makes this method much cleaner.

Comment (peterkeating) [reply Helpers/CommonMenuItemPartMigrator.cs:104]: Method is async, which means method name should be postfixed with "Async".

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        {
+            try
+            {
+                var contentItem = await _session.Query<ContentItem>()
+                    .With<ContentItemIndex>()
+                        .Where(x => x.ContentItemId == linkMenuItem.As<CommonMenuItem>().LinkTo.ContentItemIds[0])
+                    .FirstOrDefaultAsync();
+
+                return contentItem.As<AutoroutePart>().Path;
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:113]: Potential null reference exception here in multiple ways.

What happens when there is no `contentItem`? What happens when the content item doesn't have an `AutoroutePart`? These scenarios must be catered for.

File: Helpers/CommonMenuItemPartMigrator.cs (hunk 6 above)
Comment (peterkeating) [reply Helpers/CommonMenuItemPartMigrator.cs:104]: Method can be `private` too.

File: Migrations.cs [hunk 8]
```diff
@@ -1,22 +1,33 @@
...
 
 namespace Etch.OrchardCore.Menu
 {
     public class Migrations : DataMigration
     {
         private readonly IContentDefinitionManager _contentDefinitionManager;
+        private readonly IContentManager _contentManager;
+        private readonly ICommonMenuItemPartMigrator _commonMenuitemPartMigrator;
         private readonly IRecipeMigrator _recipeMigrator;
+        private readonly ISession _session;
```
Comment (peterkeating) [inline Migrations.cs:22]: `IContentManager` & `ISession` aren't used so can be removed.

Comment (peterkeating) [reply Migrations.cs:22]: Sort args alphabetically.

File: Models/CommonMenuItem.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Etch.OrchardCore.Menu.Models
+{
+    public class CommonMenuItem : ContentPart
+    {
+        public ContentPickerField LinkTo { get; set; }
+        public TextField ExternalUrl { get; set; }
+        public BooleanField OpenNewTab { get; set; }
+        public CodeField OnClick { get; set; }
+
```
Comment (peterkeating) [inline Models/CommonMenuItem.cs:13]: Remove empty line.

File: Models/CommonMenuItem.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿using Etch.OrchardCore.Fields.Code.Fields;
+using OrchardCore.ContentFields.Fields;
+using OrchardCore.ContentManagement;
+
+namespace Etch.OrchardCore.Menu.Models
+{
+    public class CommonMenuItem : ContentPart
+    {
+        public ContentPickerField LinkTo { get; set; }
```
Comment (peterkeating) [inline Models/CommonMenuItem.cs:9]: Sort properties alphabetically.

File: Startup.cs [hunk 11]
```diff
@@ -8,7 +11,9 @@ public class Startup : StartupBase
     {
         public override void ConfigureServices(IServiceCollection services)
         {
+            services.AddHttpContextAccessor();
```
Comment (peterkeating) [inline Startup.cs:14]: Why is this needed?

File: Views/MenuItem-ImageMenuItem.Admin.cshtml [hunk 12]
```diff
@@ -6,21 +6,12 @@
 @{
     ContentItem contentItem = Model.ContentItem;
     var menuItems = contentItem.As<MenuItemsListPart>();
+    var linkDestinationPart = contentItem.Content.LinkDestinationPart;
     var index = 0;
 
     var linkUrl = string.Empty;
+    linkUrl = linkDestinationPart.DestinationURL.Text.Value;
```
Comment (peterkeating) [inline Views/MenuItem-ImageMenuItem.Admin.cshtml:13]: This property should be "DestinationUrl"?

File: Helpers/CommonMenuItemPartMigrator.cs (hunk 2 above)
Comment (Mehdi-Miah) [reply Helpers/CommonMenuItemPartMigrator.cs:71]: The issue is here, when I use httpcontextaccessor. it returns to me a path such as "/studio/admin" i assume the admin is there as the migration is running in the background therefore it is returning the "admin" 

so my thinking was to remove admin from the url which would leave me with "/studio/" of which i append the contentitemId autoroute path "homepage" (in this example) which returns a url of "/studio/homepage.

Comment (peterkeating) [reply Helpers/CommonMenuItemPartMigrator.cs:71]: You shouldn't need to use the `HttpContext`.

Comment (Mehdi-Miah) [reply Helpers/CommonMenuItemPartMigrator.cs:71]: what should i be using instead?

Comment (peterkeating) [reply Helpers/CommonMenuItemPartMigrator.cs:71]: If you're trying to get the request URL prefix then you should be getting it from the [Shell Settings](https://github.com/OrchardCMS/OrchardCore/blob/5e37e29c54d44bffeeae529f8fd2cf3359734911/src/OrchardCore/OrchardCore.Abstractions/Shell/ShellSettings.cs#L61)

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+                await _contentManager.UpdateAsync(contentItem);
+                await _contentManager.PublishAsync(contentItem);
+            };
+        }
+
+        private async Task<string> ContentItemIdToUrlAsync(string contentItemId)
+        {
+            var contentItem = await _session.Query<ContentItem>()
+                .With<ContentItemIndex>()
+                    .Where(x => x.ContentItemId == contentItemId)
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:57]: Make sure we get the latest here.

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+                .FirstOrDefaultAsync();
+
+            if (contentItem == null)
+            {
+                return string.Empty;
+            }
+
+            if (contentItem.As<AutoroutePart>() != null)
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:65]: Use `Has`.

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+            {
+                return contentItem.As<AutoroutePart>().Path;
+            }
+
+            return string.Empty;
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:70]: If there is no `AutoroutePart` there is still a URL for a content item, it's something like `Content/Display/<contentItemId>`. Try creating a content type without an autoroute part and publishing a content item, that will tell you what the URL should be.

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+        {
+            foreach (var linkMenuItem in menuItemsListPart.MenuItems)
+            {
+                var linkDestinationPart = new LinkDestinationPart();
+
+                try
+                {
+                    if (linkMenuItem.As<CommonMenuItem>().ExternalUrl.Text != null)
+                    {
+                        linkDestinationPart = new LinkDestinationPart { DestinationUrl = linkMenuItem.As<CommonMenuItem>().ExternalUrl };
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:91]: You already instantiate `LinkDestinationPart` on line 85. So this can change to `linkDestinationPart.DestinationUrl = linkMenuItem.As<CommonMenuItem>().ExternalUrl`

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+                    }
+                    else
+                    {
+                        var path = "/";
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:95]: Move this logic in to `ContentItemIdToUrlAsync` so you have the following:

```
linkDestinationPart.DestinationUrl = await ContentItemIdToUrlAsync(linkMenuItem.As<CommonMenuItem>().LinkTo.ContentItemIds[0]);
```
```

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+
+        public async Task<MenuItemsListPart> MenuItemMigratorAsync(MenuItemsListPart menuItemsListPart)
+        {
+            foreach (var linkMenuItem in menuItemsListPart.MenuItems)
+            {
+                var linkDestinationPart = new LinkDestinationPart();
+
+                try
+                {
+                    if (linkMenuItem.As<CommonMenuItem>().ExternalUrl.Text != null)
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:89]: We need to cater for when the menu item doesn't have the `CommonMenuItem` otherwise this will error.

**Review by peterkeating — CHANGES_REQUESTED** [review]

File: Helpers/CommonMenuItemPartMigrator.cs (hunk 18 above)
Comment (Mehdi-Miah) [reply Helpers/CommonMenuItemPartMigrator.cs:89]: this is caught and the loop continues to run through the list of menu items whilst logging the error

Comment (peterkeating) [reply Helpers/CommonMenuItemPartMigrator.cs:89]: If we can prevent the error then we should do so. That way only actual errors will be logged. Add a check for whether the content item has the `CommonMenuItem` part, if it doesn't then skip to the next item in the loop.

File: Helpers/CommonMenuItemPartMigrator.cs (hunk 15 above)
Comment (Mehdi-Miah) [reply Helpers/CommonMenuItemPartMigrator.cs:70]: so i created a type, added only a title part to it. 

when i go to view the page the url is - `https://localhost:44358/studio/Contents/ContentItems/44ctbdvfs2nyf12mx39v2qg9cd`

so safe to assume setting path as `/studio/Contents/ContentItems/contentItemId`

i can add '"Contents/ContentItems/contentItemId"' at the end of the path variable which adds the url prefix

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+                .FirstOrDefaultAsync();
+
+            if (contentItem == null)
+            {
+                return path;
+            }
+
+            if (contentItem.Has<AutoroutePart>())
+            {
+                return path += contentItem.As<AutoroutePart>().Path.ToString();
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:69]: This should be `return  $"{path}{contentItem.As<AutoroutePart>().Path}"`. Think SonarCloud picked this up too.

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+            }
+
+            return path += $"Contents/ContentItems/{contentItemId}";
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:72]: Same issue as above.

File: Helpers/CommonMenuItemPartMigrator.cs [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+
+        private async Task<IEnumerable<ContentItem>> FetchMenuItemsAsync()
+        {
+            return await _session.Query<ContentItem>()
+                .With<ContentItemIndex>()
+                    .Where(x => x.Latest && x.ContentType == "Menu")
+                .ListAsync();
+        }
+
+        public async Task<MenuItemsListPart> MenuItemMigratorAsync(MenuItemsListPart menuItemsListPart)
```
Comment (peterkeating) [inline Helpers/CommonMenuItemPartMigrator.cs:83]: Doesn't need to be public. Change method name to "MigrateMenuItemsAsync", gives more of an impression of what it's doing to do.

**Review by peterkeating — CHANGES_REQUESTED** [review]

**Review by peterkeating — APPROVED** [review]
